atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/OracleShrinkExecutor.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.keyvalue.dbkvs.impl;
+
+import java.sql.SQLException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Stopwatch;
+import com.palantir.atlasdb.keyvalue.api.TableReference;
+import com.palantir.atlasdb.keyvalue.dbkvs.OracleDdlConfig;
+import com.palantir.atlasdb.keyvalue.dbkvs.OracleTableNameGetter;
+import com.palantir.atlasdb.keyvalue.impl.TableMappingNotFoundException;
+import com.palantir.atlasdb.logging.LoggingArgs;
+import com.palantir.exception.PalantirSqlException;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.nexus.db.sql.SqlConnection;
+
+public class OracleShrinkExecutor {
+    private static final Logger log = LoggerFactory.getLogger(OracleShrinkExecutor.class);
+
+    private final SqlConnectionSupplier connectionPool;
+    private final ExecutorService executorService;
+    private final OracleTableNameGetter oracleTableNameGetter;
+    private OracleDdlConfig oracleDdlConfig;
+
+    private static Future<Boolean> previousShrinkFuture;
+
+    public OracleShrinkExecutor(
+            SqlConnectionSupplier connectionPool,
+            ExecutorService executorService,
+            OracleTableNameGetter oracleTableNameGetter,
+            OracleDdlConfig oracleDdlConfig) {
+        this.connectionPool = connectionPool;
+        this.executorService = executorService;
+        this.oracleTableNameGetter = oracleTableNameGetter;
+        this.oracleDdlConfig = oracleDdlConfig;
+    }
+
+    public void shrinkAsync(TableReference tableRef) {
+        final long secondsToWaitBeforeShrink = getSecondsToWaitBeforeShrink();
+        previousShrinkFuture = executorService.submit(
+                () -> shrinkCompactFollowedByShrink(tableRef, secondsToWaitBeforeShrink));
+
+    }
+
+    private long getSecondsToWaitBeforeShrink() {
```

Comment:
This method has a lot of nesting - let's cut down on that.

We could have a boolean-valued function `shouldWaitLonger()` (which utilise early returns to have at most one level of nesting), and then this becomes:
```
return shouldWaitLonger() ? oracleDdlConfig.shrinkPauseSeconds() + oracleDdlConfig.shrinkPauseOnFailureSeconds()
                          : oracleDdlConfig.shrinkPauseSeconds();
```

Reply:
Done.